Add a cloudscale_volume resource

davidhalter opened the pull request:
I have added a cloudscale_volume resource to make it possible to manage volumes in terraform.

I have added documentation, tests and I don't think there is more to do. Please let me know what you think. Thanks for looking at it!

Once this is merged, it would be really nice if we could release it :)

~ Dave

davidhalter committed:
Upgrade cloudscale-go-sdk

davidhalter committed:
Add an initial untested version for a new resource cloudscale_volume

davidhalter committed:
Sweep all objects in case of errors instead of just one.

davidhalter committed:
Add some cloudscale_volume resource tests

davidhalter committed:
Add a reattach test

davidhalter committed:
Add a volume documentation

davidhalter committed:
go fmt

ferdinandhuebner commented:
I would really like to see this merged. 

Is there anything I can help with to move this PR forward?

davidhalter commented:
I have to wait for HashiCorp employees to take a look at it. If they don't react soon, I'll try their slack.

davidhalter committed:
Fix an issue in make vet

davidhalter committed:
Use the proper cloudscale-go-sdk for volumes

davidhalter commented:
I tried to get them to review it, but no answer, yet. I will just write again after a while.

davidhalter commented:
Since we're allowed to merge ourselves now, would you @ferdinandhuebner be able to review my code so we can merge it?

I think you probably know golang a bit better than my coworkers at cloudscale.ch, so I would prefer you reviewing my code!

ferdinandhuebner commented:
Sure, I'll take a look at it next week.

ferdinandhuebner reviewed (commented):
Before I take a closer look at the code, I have a question or observation about the way the volume resource is currently modelled and what the implications are, especially when it comes to deleting resources, servers in particular. 

At the moment, the volume resource is aligned with the cloudscale.ch API when it comes to attaching volumes to a server, i.e. the server where the volume is going to be attached to is specified on the volume resource. That makes sense.

For other providers, e.g. AWS, GCP, OpenStack or DigitalOcean, it's the other way around and you usually add the volume attachment to the server resource. 
With the OpenStack and DigitalOcean providers, there is a separate attachment resource, e.g. [compute_volume_attach_v2](https://www.terraform.io/docs/providers/openstack/r/compute_volume_attach_v2.html) that you could use to attach a volume to a server.

I'm not sure why, but I would have initially expected that the volume attachment happens on the server resource and the volume is just a thing that lies around waiting to be attached or destroyed.

I don't necessarily suggest that we use the same approach as the other providers, but I would like to ask what would happen if a server gets destroyed that has a volume attached to it. 

Does the volume get deleted like it is currently happening if I were to delete the server through the API? Do we want this behaviour in this case?

I think a separate volume attachment resource seems unnecessary at first, but it might help avoid accidental or unexpected deletion of a volume that is attached to a server. 

Consider this use-case: I create a server and a volume. The volume holds important data that I want to keep. I want to recreate the server for some reason. 
As a user, I would expect that I can just `terraform taint` the server and run `terraform apply`  to force re-creation. I don't know enough about resource dependencies in terraform, but would an attachment resource be the thing that would prevent the volume from being deleted by the cloudscale.ch API?

davidhalter commented:
Good question. We know that this is still an issue with our API - I will discuss this internally. Thanks!

ferdinandhuebner commented:
Thank you. Personally, I would think about making the current behaviour the concern of the GUI and not the API. On the GUI, you can ask the user for consent, e.g. "Do you want to delete volumes attached to this server?". It could then detach and delete the volumes and delete the server.

I thought about the attachment resource again and I don't see a real benefit for it. Whatever you decide on the API behaviour regarding server deletion, we could work around the current API behaviour on the terraform provider and detach all volumes prior to deleting a server.

I'll take a look at the code now.

ferdinandhuebner commented on website/docs/r/volume.html.markdown:
```diff
@@ -0,0 +1,46 @@
+}
+
+# Add a Volume to web-worker01
+resource "cloudscale_volume" "web-worker01-vip" {
```
Can you rename the resource to e.g. `web-worker01-volume` like in the `index.html.markdown`? The current name seems to be a copy/paste from the floating IP documentation.

ferdinandhuebner commented on website/docs/r/volume.html.markdown:
```diff
@@ -0,0 +1,46 @@
+The following arguments are supported when creating/changing Volumes:
+
+* `name` - (Required) Name of the new volume.
+* `size_gb` - (Required) The volume size in GB. Valid values are multiples of 50 for type "ssd" and multiples of 100 for type "bulk".
```
Update the valid increments for SSDs to 1 GB instead of 50 GB

ferdinandhuebner commented on website/docs/r/volume.html.markdown:
```diff
@@ -0,0 +1,46 @@
+
+The following arguments are supported when creating/changing Volumes:
+
+* `name` - (Required) Name of the new volume.
```
Is there a restriction on the name? Can it be an arbitrary length? Is the name limited to a certain set of characters?

davidhalter replied:
It can be arbitrary. The length is limited, but not to something "small", so I would vote for leaving it like that. The API documentation is a bit more precise for quite a few things (we've chosen to do it like that when contributing `cloudscale_server`). It's valid criticism, though.

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume_test.go:
```diff
@@ -0,0 +1,270 @@
+
+		// Try to find the volume
+		_, err := client.Volumes.Get(context.Background(), id)
+		if err != nil && !strings.Contains(err.Error(), "Not found") {
```
Could you use the HTTP status code here instead of checking if the error message contains a string?

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume_test.go:
```diff
@@ -0,0 +1,270 @@
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "0"),
```
Not sure if it's necessary, but you could add an additional check against the cloudscale.ch API and check if the volume is indeed detached.

davidhalter replied:
I added additional checks against the cloudscale API in all places where volume existence is checked.

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume_test.go:
```diff
@@ -0,0 +1,270 @@
+func assertVolumeAttached(server *cloudscale.Server, volume *cloudscale.Volume) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if server.UUID != (*volume.ServerUUIDs)[0] {
+			return fmt.Errorf("Server not properly attached to %s: %v", server.UUID, volume.ServerUUIDs)
```
Did you mean "volume not properly attached to %s: %v"? Would it make sense to rephrase it to something like "volume is not attached to %s but to %v"?

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume.go:
```diff
@@ -0,0 +1,177 @@
+
+	err := d.Set("server_uuids", volume.ServerUUIDs)
+	if err != nil {
+		log.Printf("[DEBUG] Error setting ssh_fingerprins attribute: %#v, error: %#v", volume.ServerUUIDs, err)
```
did you mean something like "error setting server_uuids attribute..."?

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume.go:
```diff
@@ -0,0 +1,177 @@
+
+	volume, err := client.Volumes.Get(context.Background(), d.Id())
+	if err != nil {
+		if err.Error() == "detail: Not Found." {
```
Can you use the HTTP status code?

ferdinandhuebner commented on cloudscale/resource_cloudscale_volume.go:
```diff
@@ -0,0 +1,177 @@
+	log.Printf("[INFO] Deleting Volume: %s", d.Id())
+	err := client.Volumes.Delete(context.Background(), id)
+
+	if err != nil && strings.Contains(err.Error(), "Not found") {
```
can you use the HTTP status code?

ferdinandhuebner reviewed (changes requested):
I think the code looks goode, apart from some minor suggestions.

davidhalter commented:
Hi Ferdinand

We finally decided to go with the behavior you initially suggested. Volumes will not be deleted in the API when a server is deleted.

Of course it might take us some time to implement it, because it might also have consequences for our GUI. Once that's done I will incorporate your feedback (good finds BTW) and we will merge this.

davidhalter committed:
Some fixes for the cloudscale_volume changes

See https://github.com/terraform-providers/terraform-provider-cloudscale/pull/5

davidhalter committed:
Check HTTP 404 with the status code not with a stupid error message

davidhalter committed:
Merge branch 'master' into volume

davidhalter committed:
Go fmt

davidhalter committed:
Remove an unused import

davidhalter committed:
Some cleanups of the acceptance tests

davidhalter commented:
@ferdinandhuebner As we wrote in an email to you, we have changed our API to not delete volumes when servers are deleted. This is very beneficial for the terraform implementation and makes sense in general IMO. Thanks for bringing it up.

I fixed all the things you proposed (except one that I commented) and would like to merge if possible.

ferdinandhuebner commented:
Looks good to me
